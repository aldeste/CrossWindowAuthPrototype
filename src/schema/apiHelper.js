/* @flow */
import DataLoader from "dataloader";
import chalk from "chalk";
import { Person as dbPerson, Planet as dbPlanet } from "../data";
import User, { type Viewer } from "../models/User";
import Planet from "../models/Planet";

import { type GraphQLObjectType } from "graphql";

export type Loaders = {
  Person: Function,
  Planet: Function
};

export type DataLoaders = {
  Person: DataLoader<*, *>,
  Planet: DataLoader<*, *>
};

/**
 * This resolves a query to each id, batching the array into a
 * single fetch. this function is memoized as a response to DataLoader,
 * meaning the same ID won't be fetched twice in the same query.
 */
export function getResolve(
  type: Class<*>,
  ids: Array<string>,
  options: Object = {}
): Promise<*> {
  return Promise.all(
    ids.map(id => {
      // We console log each request to ilustrate how dataloader works,
      // If the same reqouse is requisted twice durning the same request
      // it'll still only consol log once, since the result is memoized.
      console.log(
        chalk.bold.magenta(
          "Fetched",
          chalk.yellow.italic(type.name),
          "with id",
          chalk.yellow.italic(id),
          "from database"
        )
      );
      return type.scope("withIds").findById(id, options);
    })
  );
}

/**
 * A factory to create DataLoader memoized database calls. Best
 * practice is to create each per request, this will help if
 * same data is fetched multiple times durning the same request.
 */
export function createLoaders(Viewer: ?Viewer): Loaders {
  const loaders: DataLoaders = {
    Person: new DataLoader(ids => getResolve(dbPerson, ids)),
    Planet: new DataLoader(ids => getResolve(dbPlanet, ids))
  };

  return {
    Person: (id: string): Promise<?User> =>
      User.gen(Viewer || null, id, loaders),
    PersonConnection: ids => User.genMany(Viewer || null, ids, loaders),
    Planet: (id: string): Promise<?Object> =>
      Planet.gen(Viewer || null, id, loaders)
  };
}

/**
 * Given a type and ID, get the object with the ID.
 */
export async function getObjectFromTypeAndId(
  type: Object,
  id: string,
  context?: Object
): Promise<Object> {
  const Loaders = (context && context.loaders) || createLoaders();
  return await Loaders[type.name](id);
}

/**
 * Given a type, fetch all objects
 */
export async function getObjectsByType(
  type: GraphQLObjectType,
  args?: ?Object,
  context: Object,
  ids: Array<string>
): Promise<Object> {
  const Loaders = (context && context.loaders) || createLoaders();
  const objects = await Loaders[type.name + "Connection"](ids);

  return { objects, currentCount: objects.length };
}
