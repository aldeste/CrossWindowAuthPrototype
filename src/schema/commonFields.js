// @flow

/**
 * This file contains fields used by each GraphQL type. Instead
 * of defining them multiple times, they are defined here.
 */

import { GraphQLString, type GraphQLFieldConfig } from "graphql";

export function createdField(): GraphQLFieldConfig<*, *> {
  return {
    type: GraphQLString,
    description:
      "The ISO 8601 date format of the time that this resource was created.",
    resolve: current => current.created
  };
}

export function editedField(): GraphQLFieldConfig<*, *> {
  return {
    type: GraphQLString,
    description:
      "The ISO 8601 date format of the time that this resource was edited.",
    resolve: current => current.edited
  };
}
