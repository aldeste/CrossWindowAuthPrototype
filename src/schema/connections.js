/* @flow */
import {
  connectionFromArray,
  connectionArgs,
  connectionDefinitions
} from "graphql-relay";
import { getObjectsByType } from "./apiHelper";
import {
  type GraphQLObjectType,
  type GraphQLFieldConfig,
  GraphQLInt
} from "graphql";

export function connectionFromType(
  name: string,
  type: GraphQLObjectType,
  idField: string
): GraphQLFieldConfig<*, *> {
  const { connectionType } = connectionDefinitions({
    name,
    nodeType: type,
    resolveNode: edge => edge.node,
    connectionFields: () => ({
      currentCount: {
        type: GraphQLInt,
        resolve: conn => conn.currentCount,
        description: `A count of the total number of objects in this connection, ignoring pagination.
This allows a client to fetch the first five objects by passing "5" as the
argument to "first", then fetch the total count so it could display "5 of 83",
for example.`
      }
    })
  });
  return {
    type: connectionType,
    args: connectionArgs,
    resolve: async (currentType: Object, args: Object, ctx: Object) => {
      const { objects, currentCount } = await getObjectsByType(
        type,
        args,
        ctx,
        currentType[idField].map(x => x.id)
      );
      return {
        ...connectionFromArray(objects, args),
        currentCount
      };
    }
  };
}
