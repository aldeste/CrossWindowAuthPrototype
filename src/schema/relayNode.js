// @flow
import { fromGlobalId, nodeDefinitions } from "graphql-relay";

import PersonType from "./types/personType";
import PlanetType from "./types/planetType";

import { getObjectFromTypeAndId } from "./apiHelper";

export function idFetcher(
  globalId: string,
  context: Object
): Promise<Object> | null {
  const { type, id } = fromGlobalId(globalId);
  const resolve = (type: Object) => getObjectFromTypeAndId(type, id, context);
  if (type === "people") {
    return resolve(PersonType);
  }
  if (type === "planets") {
    return resolve(PlanetType);
  }
  return null;
}

export function typeResolver(object: { GraphQLType?: Object }): Object | null {
  if (object.GraphQLType === PersonType.name) {
    return PersonType;
  }
  if (object.GraphQLType === PlanetType.name) {
    return PlanetType;
  }
  return null;
}

const { nodeInterface, nodeField } = nodeDefinitions(idFetcher, typeResolver);

export { nodeInterface, nodeField };
