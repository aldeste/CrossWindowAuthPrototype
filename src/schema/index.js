// @flow
import { GraphQLID, GraphQLObjectType, GraphQLSchema } from "graphql";
import { fromGlobalId } from "graphql-relay";

import { nodeField } from "./relayNode";
import personType from "./types/personType";
import { getObjectFromTypeAndId } from "./apiHelper";

type arguments = { personId?: ?string, id?: ?string };
type globalIdObejct = { type?: ?string, id?: ?string };

const Root = new GraphQLObjectType({
  name: "Root",
  fields: () => ({
    viewer: {
      type: personType,
      args: {
        id: { type: GraphQLID },
        personId: { type: GraphQLID }
      },
      resolve(_: *, args: arguments, viewer: Object): Object {
        const getter = (id: string) =>
          getObjectFromTypeAndId(personType, id, viewer);
        const { personId, id: globalId } = args;

        if (typeof personId === "string") {
          return getter(personId);
        }

        if (typeof globalId === "string") {
          const { id, type }: globalIdObejct = fromGlobalId(globalId);
          if (!id || id === "" || type === "") {
            throw new Error(`No valid ID extracted from ${globalId}`);
          }

          return getter(id);
        }
        throw new Error(`You must provide id or personId`);
      }
    },
    node: nodeField
  })
});

export default new GraphQLSchema({ query: Root });
