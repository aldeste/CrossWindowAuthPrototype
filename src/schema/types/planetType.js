/* @flow */
import {
  GraphQLFloat,
  GraphQLInt,
  GraphQLList,
  GraphQLObjectType,
  GraphQLString
} from "graphql";
import { globalIdField } from "graphql-relay";
import { createdField, editedField } from "../commonFields";
import { nodeInterface } from "../relayNode";
import { connectionFromType } from "../connections";

// import FilmType from "./film";
import PersonType from "./personType";

const PlanetType = new GraphQLObjectType({
  name: "Planet",
  description: `A large mass, planet or planetoid in the Star Wars Universe, at the time of 0 ABY.`,
  fields: () => ({
    name: {
      type: GraphQLString,
      description: "The name of this planet.",
      resolve: planet => planet.name
    },
    diameter: {
      type: GraphQLInt,
      description: "The diameter of this planet in kilometers.",
      resolve: planet => planet.diameter
    },
    rotationPeriod: {
      type: GraphQLInt,
      resolve: planet => planet.rotation_period,
      description: `The number of standard hours it takes for this planet to complete a single rotation on its axis.`
    },
    orbitalPeriod: {
      type: GraphQLInt,
      resolve: planet => planet.orbital_period,
      description: `The number of standard days it takes for this planet to complete a single orbit of its local star.`
    },
    gravity: {
      type: GraphQLString,
      description: `A number denoting the gravity of this planet, where "1" is normal or 1 standard G. "2" is twice or 2 standard Gs. "0.5" is half or 0.5 standard Gs.`,
      resolve: planet => planet.gravity
    },
    population: {
      type: GraphQLInt,
      description: "The average population of sentient beings inhabiting this planet.",
      resolve: planet => planet.population
    },
    climates: {
      type: new GraphQLList(GraphQLString),
      resolve: planet => {
        return planet.climates && planet.climates.split(",").map(s => s.trim());
      },
      description: "The climates of this planet."
    },
    terrains: {
      type: new GraphQLList(GraphQLString),
      resolve: planet => {
        return planet.terrains && planet.terrains.split(",").map(s => s.trim());
      },
      description: "The terrains of this planet."
    },
    surfaceWater: {
      type: GraphQLFloat,
      resolve: planet => planet.surface_water,
      description: `The percentage of the planet surface that is naturally occuring water or bodies of water.`
    },
    residentConnection: {
      ...connectionFromType("PlanetResidents", PersonType, "residents"),
      description: "People from this planet"
    },
    // filmConnection: connectionFromUrls("PlanetFilms", "films", FilmType),
    created: createdField(),
    edited: editedField(),
    id: globalIdField("planets")
  }),
  interfaces: () => [nodeInterface]
});
export default PlanetType;
