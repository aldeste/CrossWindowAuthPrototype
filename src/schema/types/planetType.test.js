import schema from "../";
import planetType from "./planetType";
import { graphql } from "graphql";
import { createLoaders } from "../apiHelper";

beforeEach(async () => {
  console.log = jest.fn();
  await require("../../database").initializeDatabase();
});

describe("planetType is a GraphQL type", () => {
  const rootValue = {};
  const context = { loaders: createLoaders() };

  it("Has the name Planet", () => {
    expect(planetType.name).toBe("Planet");
  });

  it("should return on every field", async () => {
    const query = `query Test {
      person(personId: 1) {
        homeworld {
          name
          diameter
          rotationPeriod
          orbitalPeriod
          gravity
          population
          climates
          terrains
          surfaceWater
          residentConnection {
            edges {
              node {
                name
              }
            }
          }
          id
        }
      }
    }`;

    const { data } = await graphql(schema, query, rootValue, context);
    expect(data).toMatchSnapshot();
  });
});
