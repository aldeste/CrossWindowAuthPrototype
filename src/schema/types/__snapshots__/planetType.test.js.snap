// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`planetType is a GraphQL type should return on every field 1`] = `
Object {
  "person": Object {
    "homeworld": Object {
      "climates": Array [
        "arid",
      ],
      "diameter": 10465,
      "gravity": "1 standard",
      "id": "cGxhbmV0czo2MA==",
      "name": "Tatooine",
      "orbitalPeriod": null,
      "population": 200000,
      "residentConnection": Object {
        "edges": Array [
          Object {
            "node": Object {
              "name": "Luke Skywalker",
            },
          },
          Object {
            "node": Object {
              "name": "C-3PO",
            },
          },
          Object {
            "node": Object {
              "name": "Darth Vader",
            },
          },
          Object {
            "node": Object {
              "name": "Owen Lars",
            },
          },
          Object {
            "node": Object {
              "name": "Beru Whitesun lars",
            },
          },
          Object {
            "node": Object {
              "name": "R5-D4",
            },
          },
          Object {
            "node": Object {
              "name": "Biggs Darklighter",
            },
          },
          Object {
            "node": Object {
              "name": "Anakin Skywalker",
            },
          },
          Object {
            "node": Object {
              "name": "Shmi Skywalker",
            },
          },
          Object {
            "node": Object {
              "name": "Cliegg Lars",
            },
          },
        ],
      },
      "rotationPeriod": null,
      "surfaceWater": null,
      "terrains": Array [
        "desert",
      ],
    },
  },
}
`;
